feat(security): integrating consul token for edgex-sys-mgmt and app-services

jim-wang-intel opened the pull request:
- add edgex-sys-mgmt and app-service-rule-engine into security yaml for them to be waiting on consul-bootstrapper

Closes: #52

Signed-off-by: Jim Wang <yutsung.jim.wang@intel.com>

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/edgex-examples/blob/master/.github/CONTRIBUTING.md.

## What is the current behavior?
<!-- Please describe the current behavior and link to a relevant issue. -->
no waiting mechanism for both services of `system-management` and `app-service-rule-engine` yet

## Issue Number: #52 


## What is the new behavior?
Add waiting mechanism for both services of `system-management` and `app-service-rule-engine`

## Does this PR introduce a breaking change?
<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

- [ ] Yes
- [x ] No

## New Imports
Are there any new imports or modules? If so, what are they used for and why?
N/A

## Specific Instructions
Are there any specific instructions or things that should be known prior to reviewing?


## Other information

lenny-intel commented on compose-builder/add-security.yml:
```diff
@@ -329,3 +329,38 @@ services:
+      # for this one doesn't need the secretstore
+      EDGEX_SECURITY_SECRET_STORE: "false"
+      # service key name is always lower cases
+      SERVICE_CONFIGACCESSTOKENFILE: /tmp/edgex/secrets/appservice-rules-engine/consul-token
```
Please make update the ASC profiles for making it lower case there so we don't need these overrides. All the profiles will need updating.

jim-wang-intel replied:
ok

lenny-intel replied:
First see my comment on the PR with overrides for snap. I am suggesting a change in direction... :-)

lenny-intel reviewed (changes requested):
Need to also update `add-asc-http-export-secure.yml` and `add-asc-mqtt-export-secure.yml'

jim-wang-intel commented:
> Need to also update `add-asc-http-export-secure.yml` and `add-asc-mqtt-export-secure.yml'

arh, good catch, forget about those two ;-)

lenny-intel commented on compose-builder/add-asc-http-export-secure.yml:
```diff
@@ -26,25 +26,25 @@ services:
       - 127.0.0.1:48102:48102/tcp
-    container_name: app-service-http-export-secrets
-    hostname: app-service-http-export-secrets
+    container_name: app-http-export-secrets
```
I was considering this can, but need to make sure this doesn't impact nay TAF tests.

lenny-intel replied:
Also not sure why we need to have non-secure and separate secure instance of ASC running in TAF secure compose file. With secure Consul they both need to be secure... I think we need to rethink this and just have `add-asc-http-export-secure.yml` extend `add-asc-http-export.yml` to add security stuff, rather than be a separate instance... @cherrycl , @cloudxxx8 thoughts?

lenny-intel replied:
> IMO, yes and address any TAF tests that depend on it.

Ok, maybe not. We don't do this with the core/support services... I.e. `core data` is simply `data`, but we do have the `hostname` & `container` name match the service key as you have done here.

cloudxxx8 replied:
> Also not sure why we need to have non-secure and separate secure instance of ASC running in TAF secure compose file. With secure Consul they both need to be secure... I think we need to rethink this and just have `add-asc-http-export-secure.yml` extend `add-asc-http-export.yml` to add security stuff, rather than be a separate instance... @cherrycl , @cloudxxx8 thoughts?

Yes, I agree with @lenny-intel 's suggestion.  It could make the test cleaner.  I saw the PR has addressed this concern.  Thanks!

cloudxxx8 replied:
Oh, this is not addressed in taf folder.  I will discuss with @cherrycl and open another PR to address this.
Let's discuss in https://github.com/edgexfoundry/edgex-compose/issues/59

jim-wang-intel replied:
updated the compose files to reflect this.  the -secure one now extends from the base version.

lenny-intel commented on compose-builder/add-asc-http-export-secure.yml:
```diff
@@ -55,3 +55,7 @@ services:
       - no-new-privileges:true
     user: "${EDGEX_USER}:${EDGEX_GROUP}"
+
+  consul:
```
Move this to the top with `secretstore-setup`

jim-wang-intel replied:
fixed

lenny-intel commented on compose-builder/add-asc-mqtt-export-secure.yml:
```diff
@@ -51,7 +51,11 @@ services:
       - no-new-privileges:true
     user: "${EDGEX_USER}:${EDGEX_GROUP}"
+
+  consul:
```
Move to top with `secretstore-setup`

jim-wang-intel replied:
fixed

lenny-intel commented on compose-builder/add-security.yml:
```diff
@@ -75,13 +75,13 @@ services:
       # uncomment and modify the following line to add additional registry ACL roles on the fly
       # the list is comma-separated service keys for these services
-      #ADD_REGISTRY_ACL_ROLES: edgex-app-service-configurable-rules, device-virtual
+      #ADD_REGISTRY_ACL_ROLES: appservice-sample,appservice-rules-engine-redis, appservice-rules-engine-mqtt, appservice-push-to-core
```
change to `app-sample, app-rules-engine-redis, app-rules-engine-mqtt, app-push-to-core`

jim-wang-intel replied:
fixed

lenny-intel commented on compose-builder/add-security.yml:
```diff
@@ -329,3 +329,33 @@ services:
+
+  # this is to make sure the service is started after security-bootstrapper process is done
+  # because it needs to await Consul roles to be created
+  app-service-rules:
```
change to app-rules-engine to be consistant?

lenny-intel replied:
Nope, please ignore.

lenny-intel commented on compose-builder/add-taf-app-services.yml:
```diff
@@ -20,14 +20,14 @@ services:
       - 48105:48105/tcp
-    container_name: app-service-functional-tests
-    hostname: app-service-functional-tests
+    container_name: app-functional-tests
```
Also make service names match the service key?

lenny-intel reviewed (dismissed):
LGTM

lenny-intel reviewed (changes requested):
The makefile needs to be change to remove `asc-http-s` and  `asc-mqtt-s` and make the use of the secure extensions implicit when `asc-http` or `asc-mqtt` are used without `no-secty`, i.e. in secure mode. The TAF generation will also need to be updated.

jim-wang-intel commented:
> The makefile needs to be change to remove `asc-http-s` and `asc-mqtt-s` and make the use of the secure extensions implicit when `asc-http` or `asc-mqtt` are used without `no-secty`, i.e. in secure mode. The TAF generation will also need to be updated.

Updated and removed `asc-x-s` makefile targets

lenny-intel commented on taf/docker-compose-taf-perf-arm64.yml:
```diff
@@ -50,13 +50,13 @@ services:
       EDGEX_SECURITY_SECRET_STORE: "false"
       REGISTRY_HOST: edgex-core-consul
-      SERVICE_HOST: app-service-mqtt-export
+      SERVICE_HOST: app-mqtt-export
```
Looks like the security portions are missing here.

jim-wang-intel replied:
good catch, i missed it!

jim-wang-intel replied:
fixed

lenny-intel commented on taf/docker-compose-taf.yml:
```diff
@@ -29,7 +29,7 @@ networks:
 services:
   app-service-functional-tests:
-    container_name: app-service-functional-tests
+    container_name: app-functional-tests
```
Looks like we need to figure out how to add the security settings for app-service-functional-tests

jim-wang-intel replied:
if `app-service-functional-test` would be tested in both modes, i.e., with security and no-security modes, then probably we would do the similar way like `asc-x` with base and extending it with -secure mode.

@cherrycl  is TAF testing treating this `app-functional-tests` the same way as `asc-x`?

lenny-intel replied:
It is only used in TAF, so I don't want to clutter those files. Just need `add-taf-app-services-secure.yml``  & `add-taf-device-services-secure.yml`

jim-wang-intel replied:
device services are out of scope for this PR; we'll deal with devices services in different PR

lenny-intel replied:
Ok.

jim-wang-intel replied:
added

jim-wang-intel commented on compose-builder/Makefile:
```diff
@@ -99,26 +96,32 @@ endif
+		else
+			TOKEN_LIST:=$(TOKEN_LIST),app-mqtt-export
+		endif
+		# when there does NOT ask for no secure mode (no-secty) explicitly,
```
fixed

jim-wang-intel commented on compose-builder/Makefile:
```diff
@@ -99,26 +96,32 @@ endif
+		else
+			TOKEN_LIST:=$(TOKEN_LIST),app-http-export
+		endif
+		# when there does NOT ask for no secure mode (no-secty) explicitly,
```
fixed

lenny-intel commented on taf/docker-compose-taf.yml:
```diff
@@ -1004,10 +989,29 @@ services:
       DATABASES_PRIMARY_HOST: edgex-redis
       DATABASE_HOST: edgex-redis
       EDGEX_PROFILE: mqtt-export
```
We have a conflict with service keys since this service is using the same profile as the standard app-mqtt-export. We need to use the `EDGEX_SERVICE_KEY` environment variable to override the service key for this instance. Something like `EDGEX_SERVICE_KEY: scalability-app-mqtt-export` ? Then this key needs to be added to the tokens list.

jim-wang-intel replied:
I didn't find `EDGEX_SERVICE_KEY` env in the app-service-mqtt-export's configuration, where does it come from?
https://github.com/edgexfoundry/app-service-configurable/blob/master/res/mqtt-export/configuration.toml

lenny-intel replied:
It isn't a config override. It is a specific override of the service key that the SDK implements.

jim-wang-intel replied:
oh i see

lenny-intel reviewed (changes requested):
Getting very close...

jim-wang-intel commented:
> This need to be in the base TAF compose since it applies to both secure and non-secure.

i don't quite understand: in no-security mode, do we still need the "service-key" for that? before that it doesn't have that either.

lenny-intel commented:
> > This need to be in the base TAF compose since it applies to both secure and non-secure.
> 
> i don't quite understand: in no-security mode, do we still need the "service-key" for that? before that it doesn't have that either.

It is still a conflict between the two instances no mater secure or non-secure. We can't have two instances running with the same service key. The secure changes have just exposed the conflict.

jim-wang-intel commented:
> > > This need to be in the base TAF compose since it applies to both secure and non-secure.
> > 
> > 
> > i don't quite understand: in no-security mode, do we still need the "service-key" for that? before that it doesn't have that either.
> 
> It is still a conflict between the two instances no mater secure or non-secure. We can't have two instances running with the same service key. The secure changes have just exposed the conflict.

got you, so it is a hidden defect in no-security mode from the onset.

lenny-intel reviewed (approved):
LGTM

cherrycl commented:
recheck
